File: fc-python/benchmark_fcmath.py
import os
import sys
import time
from datetime import datetime
import json
import statistics
from config import BENCHMARK_ITERATIONS, BENCHMARK_RESULTS_DIR
from FCmath import function_call_playground

# Test cases for math functions
test_cases = [
    {
        "name": "basic_arithmetic",
        "prompt": "What is 5 plus 3?",
        "expected_function": "add",
        "expected_args": {"a": 5, "b": 3}
    },
    {
        "name": "subtraction",
        "prompt": "Calculate 10 minus 4",
        "expected_function": "subtract",
        "expected_args": {"a": 10, "b": 4}
    },
    {
        "name": "multiplication",
        "prompt": "Multiply 6 by 7",
        "expected_function": "multiply",
        "expected_args": {"a": 6, "b": 7}
    },
    {
        "name": "division",
        "prompt": "Divide 15 by 3",
        "expected_function": "divide",
        "expected_args": {"a": 15, "b": 3}
    },
    {
        "name": "derivative",
        "prompt": "Find the derivative of x^2",
        "expected_function": "derivative",
        "expected_args": {"expression": "x^2", "variable": "x"}
    },
    {
        "name": "integration",
        "prompt": "Integrate 2x with respect to x",
        "expected_function": "integrate",
        "expected_args": {"expression": "2x", "variable": "x"}
    },
    {
        "name": "equation_solving",
        "prompt": "Solve the equation x + 5 = 10",
        "expected_function": "solve_equation",
        "expected_args": {"equation": "x + 5 = 10", "variable": "x"}
    }
]

def run_benchmark():
    results = {
        "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "test_cases": []
    }
    
    for test_case in test_cases:
        case_results = {
            "name": test_case["name"],
            "prompt": test_case["prompt"],
            "iterations": []
        }
        
        for i in range(BENCHMARK_ITERATIONS):
            try:
                start_time = time.time()
                response = function_call_playground(test_case["prompt"])
                end_time = time.time()
                
                execution_time = end_time - start_time
                
                iteration_result = {
                    "iteration": i + 1,
                    "execution_time": execution_time,
                    "response": response,
                    "success": True
                }
                
            except Exception as e:
                iteration_result = {
                    "iteration": i + 1,
                    "execution_time": None,
                    "error": str(e),
                    "success": False
                }
            
            case_results["iterations"].append(iteration_result)
        
        # Calculate statistics for this test case
        successful_times = [r["execution_time"] for r in case_results["iterations"] if r["success"]]
        
        if successful_times:
            case_results["summary"] = {
                "avg_time": statistics.mean(successful_times),
                "min_time": min(successful_times),
                "max_time": max(successful_times),
                "median_time": statistics.median(successful_times),
                "std_dev": statistics.stdev(successful_times) if len(successful_times) > 1 else 0,
                "success_rate": len(successful_times) / BENCHMARK_ITERATIONS
            }
        else:
            case_results["summary"] = {
                "success_rate": 0
            }
        
        results["test_cases"].append(case_results)
    
    # Calculate overall statistics
    all_times = []
    for case in results["test_cases"]:
        if "summary" in case and "avg_time" in case["summary"]:
            all_times.append(case["summary"]["avg_time"])
    
    if all_times:
        results["summary"] = {
            "overall_avg_time": statistics.mean(all_times),
            "overall_min_time": min(all_times),
            "overall_max_time": max(all_times),
            "overall_median_time": statistics.median(all_times),
            "overall_std_dev": statistics.stdev(all_times) if len(all_times) > 1 else 0
        }
    
    # Save results
    os.makedirs(BENCHMARK_RESULTS_DIR, exist_ok=True)
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    results_file = os.path.join(BENCHMARK_RESULTS_DIR, f"math_benchmark_{timestamp}.json")
    
    with open(results_file, "w") as f:
        json.dump(results, f, indent=4)
    
    print(f"Math benchmark results saved to: {results_file}")
    return results

if __name__ == "__main__":
    run_benchmark() 

File: fc-nodejs/benchmark_fcmath.js
import {function_call_playground_math} from "./FCmath.js";
import {BENCHMARK_ITERATIONS, BENCHMARK_RESULTS_DIR} from "./config.js";
import path from 'path';
import fs from 'fs';

const prompt = [
    "what is the square root of 16?",
    "what is the derivative of x^2 with respect to x?",
    "what is 10 plus 10?",
    "what is the integral of x with respect to x?",
    "solve x+3=6 for x"
]

const iteration_count = BENCHMARK_ITERATIONS;

const run_benchmark = async()=>{
    const results = [];
    for (let i =0; i<iteration_count;i++){
        for (const p of prompt){
            console.log(`Running iteration ${i+1} of ${iteration_count} with prompt: ${p}`);
            const start_time = Date.now();
            const result = await function_call_playground_math(p);
            console.log('Debug - Raw result:', JSON.stringify(result, null, 2));
            const end_time = Date.now();
            const time_taken = (end_time-start_time)/1000;
            results.push({
                prompt: p,
                time_taken: time_taken,
                result: result
            });
        }
    }
    return results;
}

const success_rate = (results)=>{
    const success_count = results.filter(r=>r.result.isValid).length;
    return (success_count/results.length)*100;
}

const min_time = (results)=>{
    const valid_times = results.filter(r=>r.result.isValid).map(r=>r.time_taken);
    return valid_times.length > 0 ? Math.min(...valid_times) : 0;
}

const max_time = (results)=>{
    const valid_times = results.filter(r=>r.result.isValid).map(r=>r.time_taken);
    return valid_times.length > 0 ? Math.max(...valid_times) : 0;
}

const avg_time = (results)=>{
    const valid_times = results.filter(r=>r.result.isValid).map(r=>r.time_taken);
    return valid_times.length > 0 ? valid_times.reduce((sum,t)=>sum+t,0)/valid_times.length : 0;
}

const median_time = (results)=>{
    const valid_times = results.filter(r=>r.result.isValid).map(r=>r.time_taken).sort((a,b)=>a-b);
    if (valid_times.length === 0) return 0;
    const mid = Math.floor(valid_times.length/2);
    return valid_times[mid];
}

const std_dev = (results)=>{
    const valid_times = results.filter(r=>r.result.isValid).map(r=>r.time_taken);
    if (valid_times.length === 0) return 0;
    const avg = avg_time(results);
    const variance = valid_times.reduce((sum,t)=>sum+(t-avg)**2,0)/valid_times.length;
    return Math.sqrt(variance);
}

const save_results = (results)=>{
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `fcmath_benchmark_${timestamp}.json`;
    
    // Create the directory if it doesn't exist
    if (!fs.existsSync(BENCHMARK_RESULTS_DIR)) {
        fs.mkdirSync(BENCHMARK_RESULTS_DIR, { recursive: true });
    }
    
    const filepath = path.join(BENCHMARK_RESULTS_DIR, filename);
    fs.writeFileSync(filepath, JSON.stringify(results, null, 2));
    console.log(`Results saved to: ${filepath}`);
}

export const main_math = async()=>{
    const results = await run_benchmark();
    const successRate = success_rate(results);
    const minTime = min_time(results);
    const maxTime = max_time(results);
    const avgTime = avg_time(results);
    const medianTime = median_time(results);
    const stdDev = std_dev(results);
    
    console.log(`Benchmark Results:`);
    console.log(`Success Rate: ${successRate.toFixed(2)}%`);
    console.log(`Min Time: ${minTime.toFixed(2)}s`);
    console.log(`Max Time: ${maxTime.toFixed(2)}s`);
    console.log(`Avg Time: ${avgTime.toFixed(2)}s`);
    console.log(`Median Time: ${medianTime.toFixed(2)}s`);
    console.log(`Std Dev: ${stdDev.toFixed(2)}s`);
    
    // Add detailed error reporting
    const failedResults = results.filter(r => !r.result.isValid);
    if (failedResults.length > 0) {
        console.log('\nFailed Operations:');
        failedResults.forEach(r => {
            console.log(`- Prompt: "${r.prompt}"`);
            console.log(`  Error: ${r.result.error}`);
        });
    }
    
    save_results(results);
}

main_math();



